src/crafting.cpp

```diff
@@ -24,6 +24,28 @@
 #include <math.h>    //sqrt
 #include <algorithm> //std::min
 
+enum usage {
+	use_from_map = 1,
+	use_from_player = 2,
+	use_from_both = 1 | 2,
+	use_from_none = 4,
+	cancel = 5
+};
+
+struct comp_selection {
```

Comment:
You're right. I just got side tracked yesterday wondering how to do type constraints, not finding a quick answer, and putting off the template till later.